### DISCUSSION TITLE
Implied relationships for some filtered views aren't rendered

### DISCUSSION THREAD

**Discussion by richhosek** [structurizr/java#308] [Q&A] [answered]
### Expected

When I create a filtered view, I expect to see all relationships and implied relationships that match the filter.

### Actual

For some implied relationships, the relationship does not render.

### Steps to reproduce

- Create a model that has a person and a software system.
- Add a container that has two tags "Setup,Payment"
- Inside the container add two or more components that have either the tag "Setup" or the Tag "Payment"
- Create relationships between the person and the components. For the relationships connecting the person to a component tagged "Setup", tag the relationship with the same tag, and for relationships connecting the person to a component tagged "Payment", tag the relationship with the same tag.
- create a container view for the software system, include *.
- create three filtered views:
  - one that includes "Element,Relationship"
  - one that includes only the tag "Payment"
  - one that includes only the tag "Setup"
- In the rendered views, the implied relationships between the person and the container will render for the filtered view that includes "Element,Relationship" and "Setup" but NOT "Payment"

```
workspace {

    model {
        buyer = person "Buyer" {
            tags "Person,Setup,Payment"
        }
        paymentflow = softwareSystem "Payment Flow" {
            buyerapps = container "Buyer Portal / BuyerApp / Branch (ASAP)" {
                tags "Setup,Payment"
                achprofile = component "ACH (Profile)" {
                    tags "Setup"
                }
                ccprofile = component "Credit Card (Profile)" {
                    tags "Setup"
                }
                ppprofile = component "PayPal (Profile)" {
                    tags "Setup"
                }
                achpayment = component "ACH Payment" {
                    tags "Payment"
                }
                creditpayment = component "Credit Payment" {
                    tags "Payment"
                }
                btpayment = component "Brain Tree Payment" {
                    tags "Payment"
                }
            }
        }
        buyer -> achprofile "Uses" "" "Setup"
        buyer -> ccprofile "Uses" "" "Setup"
        buyer -> ppprofile "Uses" "" "Setup"
        buyer -> achpayment "Uses" "" "Payment"
        buyer -> creditpayment "Uses" "" "Payment"
        buyer -> btpayment "Uses" "" "Payment"
    }   

    views {
        systemLandscape "Buyer"  {

            include *
        }
        container paymentflow pfview "Payment Flow" {
            include *
        }

        filtered pfview include "Payment" "pfviewpayment" 
        filtered pfview include "Setup" "pfviewsetup"
        filtered pfview include "Element,Relationship" "pfviewall"

        component buyerapps baview "Buyer Apps" {
            include *
        }
        
        filtered baview include "Payment" "baviewpayment"
        filtered baview include "Setup" "baviewsetup"   
        filtered baview include "Element,Relationship" "baviewall"

        styles {
            element "Person" {
                shape person
            }
        }
    }
    configuration {
        scope none
    }
}
```

### Version/build information

Structurizr Lite
com.structurizr.lite.StructurizrLite     :  - build: 2024.03.03 (2024-03-03T10:05:43Z)
com.structurizr.lite.StructurizrLite     :  - structurizr-java: v2.1.1
com.structurizr.lite.StructurizrLite     :  - structurizr-dsl: v2.1.1
com.structurizr.lite.StructurizrLite     :  - structurizr-import: v2.1.1
com.structurizr.lite.StructurizrLite     :  - structurizr-graphviz: v2.1.1

### Severity

Major

### Priority

I have no budget and there's no rush, please fix this for free

**Comment by richhosek** [comment]
Further investigation indicates it's related to the alphabetical order of the tag names.
The one that comes first alphabetically does NOT render its filtered implied relationships.
The one that comes second DOES...

**Comment by simonbrowndotje** [comment]
This is working as expected ... the [implied relationships feature](https://docs.structurizr.com/dsl/cookbook/implied-relationships/) creates a relationship between `buyer` and `buyerapps` from `buyer -> achprofile "Uses" "" "Setup"`. After this, no more relationships between `buyer` and `buyerapps` are created. There are several options to achieve what you're trying to do:

- Use a different implied relationships strategy.
- Explicitly define the relationship between `buyer` and `buyerapps` with both tags (this needs to be done before all other relationships are defined).
- etc

Moving to a discussion...

**Reply by richhosek** [reply]
I have to say, creating explicit relationships to compensate for the way the filtered view renders implicit relationships in filtered component views is making my container view much busier than I think you intended.

**Reply by simonbrowndotje** [reply]
If you're willing to modify the descriptions of your relationships from `Uses`, you can use the `CreateImpliedRelationshipsUnlessSameRelationshipExistsStrategy` implied relationships strategy, which will create both relationships for you:

```
workspace {

    model {

        !script groovy {
            workspace.model.setImpliedRelationshipsStrategy(new com.structurizr.model.CreateImpliedRelationshipsUnlessSameRelationshipExistsStrategy());
        }

        buyer = person "Buyer" {
            tags "Person,Setup,Payment"
        }
        paymentflow = softwareSystem "Payment Flow" {
            buyerapps = container "Buyer Portal / BuyerApp / Branch (ASAP)" {
                tags "Setup,Payment"
                achprofile = component "ACH (Profile)" {
                    tags "Setup"
                }
                ccprofile = component "Credit Card (Profile)" {
                    tags "Setup"
                }
                ppprofile = component "PayPal (Profile)" {
                    tags "Setup"
                }
                achpayment = component "ACH Payment" {
                    tags "Payment"
                }
                creditpayment = component "Credit Payment" {
                    tags "Payment"
                }
                btpayment = component "Brain Tree Payment" {
                    tags "Payment"
                }
            }
        }
        buyer -> achprofile "Uses for setup" "" "Setup"
        buyer -> ccprofile "Uses for setup" "" "Setup"
        buyer -> ppprofile "Uses for setup" "" "Setup"
        buyer -> achpayment "Uses for payment" "" "Payment"
        buyer -> creditpayment "Uses for payment" "" "Payment"
        buyer -> btpayment "Uses for payment" "" "Payment"
    }   

    views {
        systemLandscape "Buyer"  {

            include *
        }
        container paymentflow pfview "Payment Flow" {
            include *
        }

        filtered pfview include "Payment" "pfviewpayment" 
        filtered pfview include "Setup" "pfviewsetup"
        filtered pfview include "Element,Relationship" "pfviewall"

        component buyerapps baview "Buyer Apps" {
            include *
        }
        
        filtered baview include "Payment" "baviewpayment"
        filtered baview include "Setup" "baviewsetup"   
        filtered baview include "Element,Relationship" "baviewall"

        styles {
            element "Person" {
                shape person
            }
        }
    }
    configuration {
        scope none
    }
}
```

<img width="1014" alt="Screenshot 2024-06-30 at 09 47 25" src="https://github.com/structurizr/java/assets/1009874/e1d97663-6379-4368-aee4-3117a82bd0e3">

**Comment by richhosek** [comment]
I see what you're saying, but it would be nice if a filtered view didn't just not display the items, but removed them from the underlying model -- if I'm assuming how it's working correctly. 

It appears that when rendering a filtered view, you're simply including or excluding Elements and Relationships that the condition from being displayed. The underlying model is unchanged.
If the filter worked at the model level, that would solve my problem and avoid the need to create display-only relationships on top of the implied relationships.

**Reply by simonbrowndotje** [reply]
> If the filter worked at the model level, that would solve my problem and avoid the need to create display-only relationships on top of the implied relationships.

The filter *does* work at the model level - there is no relationship between `buyer` and `buyerapps` with a tag of `Payment` in the model, which is why you don't see it on the filtered view.

![image](https://github.com/structurizr/java/assets/1009874/69413bf9-50d4-439e-ae31-67fe876f5e35)

**Reply by richhosek** [reply]
But what's happening is the implied relationships are being built using the Setup tagged relationships.
So, when I do a filtered relationship to include "Setup" I **do see the implied relationships there.**
However, when I filter to include "Payment", since the implied relationship was inferred from a relationship tagged as "Setup" the filtered Payment view won't show it, since it's still using the "Setup" tagged relationships to build the implied relationships. 

What I'm wondering is if there's a way for the implied relationships to be inferred ONLY from the filtered relationships. That way, it wouldn't see the "Setup" tags, and instead the implied relationship would inherit the "Payment" tag instead.

Right now, even when I have a filtered view, when I inspect the relationships in the javascript console, the other tagged relationshipsare still there and the implied relationships have a tag that that shouldn't exist in that view. I hope that makes things clearer.

**Reply by richhosek** [reply]
For the "Setup" filtered view I would expect to see something like this:

![C4relationships-Setup](https://github.com/structurizr/java/assets/727268/a51ad483-5d42-49a0-801f-e78785598822)
 
And for the "Payment" filtered view, I would expect to see something like this:

![C4relationships-Payment](https://github.com/structurizr/java/assets/727268/3416173e-852f-4f0f-8fc5-e72d09ac7140)

**Reply by simonbrowndotje** [reply, marked as answer]
> But what's happening is the implied relationships are being built using the Setup tagged relationships.

Correct, because they are the first to be defined in your workspace, and the [CreateImpliedRelationshipsUnlessAnyRelationshipExistsStrategy](https://docs.structurizr.com/java/implied-relationships#createimpliedrelationshipsunlessanyrelationshipexistsstrategy) prevents more from being created, which is how it's designed to work. If you'd like to change this behaviour, you will need to build your own [ImpliedRelationshipsStrategy](https://github.com/structurizr/java/blob/master/structurizr-core/src/main/java/com/structurizr/model/ImpliedRelationshipsStrategy.java) implementation that sometimes adds tags to existing relationships. You may want to use [CreateImpliedRelationshipsUnlessSameRelationshipExistsStrategy](https://docs.structurizr.com/java/implied-relationships#createimpliedrelationshipsunlesssamerelationshipexistsstrategy) ([source](https://github.com/structurizr/java/blob/master/structurizr-core/src/main/java/com/structurizr/model/CreateImpliedRelationshipsUnlessSameRelationshipExistsStrategy.java)) as a starting point.
